### DISCUSSION TITLE
Is goja.Callable go routine safe?

### DISCUSSION THREAD

**Discussion by david-sackstein** [dop251/goja#452] [Q&A]
According to the [documentation](https://github.com/dop251/goja#is-it-goroutine-safe) 
> An instance of goja.Runtime can only be used by a single goroutine at a time
My questions are as follows:
If I create a `goja.Callable` using `AssertFunction()` in one go routine, 
1. May I invoke it on another goroutine?
2. May I invoke concurrently on two other goroutines?
If this is not possible, is there another way to compile a Javascript function and invoke it with arguments concurrently on other goroutines?

**Comment by david-sackstein** [comment]
Indeed [Compile](https://github.com/dop251/goja/blob/3b8a68ca89b4/runtime.go#L1276) produces a Program that can be invoked concurrently on any goroutine, but as far as I can tell, it does not allow the caller to pass in arguments at the call site.

**Reply by dop251** [reply]
The rule here is that you cannot run any javascript code within the same Runtime concurrently, and you cannot create, access or modify Objects belonging to the same Runtime concurrently. Part of it is not even a limitation of goja, it's a limitation of ECMAScript which is non-concurrent by design. To make it work you either need a synchronisation mechanism (i.e. Mutex) or an [Event Loop](https://github.com/dop251/goja_nodejs/tree/master/eventloop).

**Reply by david-sackstein** [reply]
I would prefer to avoid locking or serialization of all operations on an event loop.
I understand that I can create an AST (Program) using Compile and then pass the program to multiple go routines and use Runtime.RunProgram on that program in each go routine concurrently.
However in that case I cannot pass a Value to a function defined in the program.
Do I understand correctly, that the way to do this is to use Set("arg", value) before running the program and having the JS reference arg as a variable rather than as an argument? This is not so clean as calling a function, but certainly workable.

Also, do you think this effort is actually worth it for performance sake?
Is RunScript much slower than Runtime.RunProgram? (of course this depends on the JS code itself, but if you can give some insights, that would be great)

**Reply by dop251** [reply]
There is no such thing as "a function defined in the program", a Program is just a pre-compiled script, it does not define any functions, they are only defined when the script (or the program) is evaluated. Using RunScript and RunProgram makes absolutely no difference other than the latter does not have to do the compilation so if you run the same program multiple times you gain performance. Also setting values and passing them as arguments makes no difference in terms of concurrency.
